network/proxy.md

```diff
@@ -6,68 +6,154 @@ keywords: network, networking, proxy, client
+> require authentication information to be included in their URL, or their
+> address may expose IP-addresses or hostnames of your company's environment.
+>
+> Environment variables are stored as plain text in the container's configuration,
+> and as such can be inspected through the remote API or committed to an image
+> when using `docker commit`.
+{: .warning }
+
+The configuration becomes active after saving the file, you don't need to
+restart Docker. However, the configuration only applies to new containers and
+builds, and doesn't affect existing containers.
+
+The following table describes the available configuration parameters.
+
+| Property     | Description                                                                         |
+| :----------- | :---------------------------------------------------------------------------------- |
+| `httpProxy`  | Sets the `HTTP_PROXY` and `http_proxy` environment variables and build arguments.   |
+| `httpsProxy` | Sets the `HTTPS_PROXY` and `https_proxy` environment variables and build arguments. |
+| `ftpProxy`   | Sets the `FTP_PROXY` and `ftp_proxy` environment variables and build arguments.     |
+| `noProxy`    | Sets the `NO_PROXY` and `no_proxy` environment variables and build arguments.       |
+| `allProxy`   | Sets the `ALL_PROXY` and `all_proxy` environment variables and build arguments.     |
+
+These settings are used to configure proxy environment variables for containers
+only, and not used as proxy settings for the Docker CLI or the Docker Engine
+itself.
+Refer to the [environment variables](/engine/reference/commandline/cli/#environment-variables)
+and [configure the Docker daemon to use a proxy server](../config/daemon/systemd.md#configure-the-docker-daemon-to-use-a-proxy-server)
+sections for configuring proxy settings for the CLI and daemon.
+
+### Run containers with a proxy configuration
+
+When you run containers, proxy-related environment variables such as
+`HTTP_PROXY` reflect the values set in the configuration file.
+
+```console
+$ docker run -it alpine
+/ # echo $HTTP_PROXY
+http://192.168.1.12:3128
+```
+
+### Build with a proxy configuration
+
+When you invoke a build, proxy-related build arguments are pre-populated
+automatically, based on the proxy settings in your Docker client configuration
+file. You can access the proxy values using `ARG` instructions in your
+Dockerfile:
+
+```dockerfile
+FROM alpine
+ARG HTTPS_PROXY
+RUN apk update    # uses httpsProxy from ~/.docker/config.json
+```
+
+### Configure proxy settings per daemon
+
+The `proxies.default` key configures the proxy settings for all daemons that
+the client connects to. To configure the proxies for individual daemons,
+specify the proxy configuration under `proxies.<hostname>` in the client
+configuration file, where hostname is the address of the daemon.
+
+For example:
+
+```json
+{
+ "proxies": {
+   "https://docker-daemon1.example.com": {
```

Comment:
I _think_ this one needs to be `tcp://docker-daemon1.example.com` (but would have to verify that).

Do you think (for a follow-up?) it would be useful to show an example as well where both "default" and "other is configured?

```bash
cat ~/.docker/config.json
{
	"proxies": {
		"default": {
			"httpProxy": "http://192.168.1.12:1111"
		},
		"unix:///Users/thajeztah/.docker/run/docker.sock": {
			"httpProxy": "http://192.168.1.12:2222"
		}
	}
}
```

And we could show examples, including how it interacts with `docker context`;

```bash
docker context ls
NAME                TYPE                DESCRIPTION                               DOCKER ENDPOINT                                   KUBERNETES ENDPOINT   ORCHESTRATOR
default *           moby                Current DOCKER_HOST based configuration   unix:///var/run/docker.sock
desktop-linux       moby                                                          unix:///Users/thajeztah/.docker/run/docker.sock


# uses "default" (if set)
docker run --rm  alpine sh -c 'env | grep _PROXY'
HTTP_PROXY=http://192.168.1.12:1111

# uses proxy for "unix:///Users/thajeztah/.docker/run/docker.sock"
docker --host unix:///Users/thajeztah/.docker/run/docker.sock run --rm  alpine sh -c 'env | grep _PROXY'
HTTP_PROXY=http://192.168.1.12:2222

# also uses proxy for "unix:///Users/thajeztah/.docker/run/docker.sock"
docker --context=desktop-linux run --rm  alpine sh -c 'env | grep _PROXY'
HTTP_PROXY=http://192.168.1.12:2222
```


(and, yes, we should integrate that with `docker context` itself, and have it be part of the context's options)


☝️ I should mention that it looks like there's currently a bug in the CLI, and setting a "per-host" proxy for a `ssh://` connection doesn't work (it tries to look up using our "magic" `http://docker.example.com` hostname, which is used internally.

Reply:
That's a cool example. Let's add it in a follow-up - for now I can try to clarify the confusion with the key names, and maybe add both a default and a "named" daemon.

TBH, I find **contexts** a confusing topic overall. Maybe that's an area worth spending some time trying to clarify.